### ISSUE TITLE
Issuing multiple commands that use agent forwarding causes script not to terminate

### ISSUE THREAD

**Issue by Toddses** [mscdex/ssh2#230]
For instance, issuing two git commands causes non termination. An example:

```javascript
var client = new Client();

client.on('ready', function() {
    console.log("Client :: ready");

    client.exec('git ls-remote --heads git@github.com:Toddses/capify-press.git', function (err, stream) {
        if (err) throw err;

        stream.on('close', function (code, signal) {
            console.log('Stream :: close :: code: ' + code + ', signal: ' + signal);
        }).on('data', function (data) {
            console.log('STDOUT: ' + data);
        }).stderr.on('data', function (data) {
            console.log('STDERR: ' + data);
        });
    });
    client.exec('git clone --mirror git@github.com:Toddses/capify-press.git /var/www/html/tester', function (err, stream) {
        if (err) throw err;

        stream.on('close', function (code, signal) {
            console.log('Stream :: close :: code: ' + code + ', signal: ' + signal);
            client.end();
            cb();
        }).on('data', function (data) {
            console.log('STDOUT: ' + data);
        }).stderr.on('data', function (data) {
            console.log('STDERR: ' + data);
        });
    });
}).connect({
    host: this.host,
    port: 22,
    user: this.user,
    agent: process.env.SSH_AUTH_SOCK,
    agentForward: true
});
```

Output

```
Client :: ready
STDOUT: Initialized empty Git repository in /var/www/html/tester/

STDOUT: f45aff911d8c23976cfcc522dce8de35681d77b7	refs/heads/dev
d762344e5829e7fd01b6233decc75f01b4addccc	refs/heads/master

Stream :: close :: code: 0, signal: undefined
Stream :: close :: code: 0, signal: undefined
```

Ignore the fact that its out of order. Even if I use something like `async.series()` and issue commands that way, I still get the same issue when using `git` more than once.

**Comment by rmg** [comment]
The output from your `git clone` looks a little short.. is it possible it is just taking longer than expected?

**Comment by rmg** [comment]
Nevermind, that doesn't make any sense given the output shows 2 close events.

**Comment by rmg** [comment]
Could this be the same as #181?

**Comment by Toddses** [comment]
I'm not opening a second connection, as in issue 181. So I wouldn't say they are the same. But considering they both are related to git commands it's certainly possible they're related issues.

**Comment by Toddses** [comment]
That is a good point regarding that output on the `git clone` though. Not sure why its giving me that output, it is doing the `clone` correctly, just giving strange output.

**Comment by Toddses** [comment]
Alright I _think_ I've managed to track down part of the issue. I know it has to do `spliceOne` function being called in the `Channel` constructor. I've found that if I issue a `git` command there is a remaining `Channel` in the `_channel` array in `client` when I call `end()`. So if I call two `git` commands, there's 2 channels in the array when `end()` is executed.

All that said, if I comment out the `spliceOne` function called in the `Channel` constructor, the script terminates just fine. So I think the issue is that when multiple `git` commands are called, there is still open channels when the script terminates.

**Comment by Toddses** [comment]
I've made multiple edits to my previous comment, as I've worked out some of the details.

**Comment by Toddses** [comment]
Here's the code I've been using to debug the issue.

```javascript
var Client = require('ssh2').Client;
var async = require('async');

var client = new Client();

client.connect({
	host: 'server',
	port: 22,
	user: 'username',
	agent: process.env.SSH_AUTH_SOCK,
	agentForward: true
});

client.on('ready', function() {
	console.log("Client :: ready");

	async.series([
		function (callback) {
			client.exec('git ls-remote --heads git@github.com:Toddses/capify-press.git', function (err, stream) {
				if (err) throw err;

				stream.on('close', function (code, signal) {
					console.log('Stream :: close :: code: ' + code + ', signal: ' + signal);
					callback();
				}).on('data', function (data) {
					console.log('STDOUT:\n' + data);
				}).stderr.on('data', function (data) {
					console.log('STDERR: ' + data);
				});
			});
			console.log('result: ' + result);
		},
		function (callback) {
			var result = client.exec('git ls-remote --heads git@github.com:Toddses/capify-press.git', function (err, stream) {
				if (err) throw err;

				stream.on('close', function (code, signal) {
					console.log('Stream :: close :: code: ' + code + ', signal: ' + signal);
					callback();
				}).on('data', function (data) {
					console.log('STDOUT:\n' + data);
				}).stderr.on('data', function (data) {
					console.log('STDERR: ' + data);
				});
			});
		},
		function (callback) {
			client.exec('git ls-remote --heads git@github.com:Toddses/capify-press.git', function (err, stream) {
				if (err) throw err;

				stream.on('close', function (code, signal) {
					console.log('Stream :: close :: code: ' + code + ', signal: ' + signal);
					callback();
				}).on('data', function (data) {
					console.log('STDOUT:\n' + data);
				}).stderr.on('data', function (data) {
					console.log('STDERR: ' + data);
				});
		}
	],
	function () {
		console.log('Client :: disconnect');
		client.end();
	});
});

```

**Comment by cmp-202** [comment]
If you are needing to run two git commands one after the other you could try https://www.npmjs.com/package/ssh2shell which is a wrapper for SSH2 shell command which is able to run multiple commands sequentially in the context of the previous commands. We originally used it for deployment automation of code updates from git to live and staging sites where we also ran package managers, node service restarts, and confirming updates completed correctly with emailing full terminal output to a developer email account and more. Terminating the connections correctly is built in to ssh2shell.

**Comment by Toddses** [comment]
Doesn't look like ssh2shell supports agent forwarding. I suspect this is why you have no problem with script termination.

**Comment by cmp-202** [comment]
It doesn't need to! It is possible to create SSH connections to any number of servers using any one of a number authentercation options for each unique connection. It can jump from one server to another in a linea path or have one server providing multiple secure connections to other servers within in its network or over the internet to a completely different location. It can run a block of different commands on each server or none at all if just providing the next hop. You can use the same authentication content between servers be it password or keys or have different authenication credentials for each. And it backsout and closes all sessions correctly when finished. You can add or remove commands as they are run based on the outcome of commands and You also get a full log of all command results for each server session. Doesn't solve your issue but might provide another option still effectively using SSH2.

**Comment by mscdex** [comment]
This should be fixed as of b1233612f0. @Toddses Let me know if master works for you now too.

**Comment by Toddses** [comment]
That did indeed fix it. Thanks, Brian.
